### ISSUE TITLE
There is room for improvement of memory allocations

### ISSUE THREAD

**Issue by supercaracal** [redis-rb/redis-cluster-client#137]
#### Node module
https://github.com/redis-rb/redis-cluster-client/blob/cabda8636492e0ec6ab3b8fa2525d86811b81f50/lib/redis_client/cluster/node.rb#L210-L219

> 352668  redis-cluster-client/lib/redis_client/cluster/node.rb
> 262224  redis-cluster-client/lib/redis_client/cluster/node.rb:211

#### Command module
https://github.com/redis-rb/redis-cluster-client/blob/master/lib/redis_client/cluster/command.rb

> 124688  redis-cluster-client/lib/redis_client/cluster/command.rb

### Memory profiling with 1000 times SET and GET

Aim: less than 200KB in redis-cluster-client/lib

#### w/ pipelining

```
allocated memory by gem
-----------------------------------
   1879853  redis-client-0.9.0
    565500  redis-cluster-client/lib
    160928  other
    115083  x64/lib
```

#### w/o pipelining

```
allocated memory by gem
-----------------------------------
   2756515  redis-client-0.9.0
    584596  redis-cluster-client/lib
    160576  other
    106809  x64/lib
```

#### w/o pipelining and w/ connection pooling

```
allocated memory by gem
-----------------------------------
   2759547  redis-client-0.9.0
   2275656  connection_pool-2.3.0
    665532  redis-cluster-client/lib
    160704  other
    106809  x64/lib
```

```
allocated memory by location
-----------------------------------
    720360  connection_pool-2.3.0/lib/connection_pool.rb:61
    720360  connection_pool-2.3.0/lib/connection_pool.rb:64
    677376  redis-client-0.9.0/lib/redis_client/ruby_connection/resp3.rb:54
    546600  redis-client-0.9.0/lib/redis_client/ruby_connection/buffered_io.rb:97
    338688  redis-client-0.9.0/lib/redis_client/ruby_connection/resp3.rb:55
    336168  connection_pool-2.3.0/lib/connection_pool.rb:68
    336168  connection_pool-2.3.0/lib/connection_pool.rb:71
    262224  redis-cluster-client/lib/redis_client/cluster/node.rb:211
    201240  redis-client-0.9.0/lib/redis_client/ruby_connection/resp3.rb:95
    184704  redis-client-0.9.0/lib/redis_client/ruby_connection/resp3.rb:147
```

**Comment by supercaracal** [comment]
TL;DR

```
irb(main):001:0> require 'objspace'
=> true

irb(main):002:0> ObjectSpace.memsize_of(Array.new(16384))
=> 131112

irb(main):005:0> s = '0' * 16384
=> "00000000000000000000000000000000000...

irb(main):006:0> ObjectSpace.memsize_of(String.new(s, encoding: Encoding::BINARY, capacity: 16384))
=> 16425

irb(main):007:0> ObjectSpace.memsize_of(Array.new(256))
=> 2088
```

```
irb(main):009:0> ObjectSpace.memsize_of({a:1, b:2, c:3, d:4, e:5})
=> 168

irb(main):010:0> s = Struct.new('Foo', :a, :b, :c, :d, :e, keyword_init: true)
=> Struct::Foo(keyword_init: true)

irb(main):011:0> ObjectSpace.memsize_of(s.new(a:1, b:2, c:3, d:4, e:5))
=> 80
```
